### DISCUSSION TITLE
What the equivalent of `__fields__[key].type_` in v2?

### DISCUSSION THREAD

**Discussion by tibbe** [pydantic/pydantic#9360] [Pydantic V2]
We have some code that uses `__fields__` in v1 to e.g. get the descriminator key of a type. Here's one such function that lives in our code base:

```py
# This is similar in spirit to
# https://github.com/pydantic/pydantic/blob/938ae5848d8d7da65b0b44f5a021fdbce95dc767/pydantic/v1/utils.py#L756.
# We need to figure out the literal value of the discriminator key.
def get_discriminator_value(
    tp: type[pydantic.BaseModel], discriminator_key: str
) -> str:
    """Returns the literal value of the discriminator key for the given type.

    The discriminator value must be a literal string.

    Raises:
      ValueError: if the discriminator field is not a `Literal`.
    """
    try:
        discriminator_tp = tp.model_fields[discriminator_key].type_
    except KeyError as e:
        raise KeyError(
            f'Model {tp.__name__!r} needs a discriminator field for key '
            '{discriminator_key!r}'
        ) from e

    if not get_origin(discriminator_tp) == Literal:
        raise ValueError(
            f'Field {discriminator_key!r} of model {tp.__name__!r} needs to be '
            'a `Literal[str]`'
        )

    tp_args = get_args(discriminator_tp)
    if len(tp_args) != 1 or not isinstance(tp_args[0], str):
        raise ValueError(
            f'Field {discriminator_key!r} of model {tp.__name__!r} needs to be '
            'a `Literal[str]`'
        )

    return tp_args[0]
```

We also sometimes use `outer_type_`.

What's the equivalent of `tp.model_fields[discriminator_key].type_` and `outer_type_` in v2?

**Comment by tibbe** [comment]
I have to ping you explicitly @samuelcolvin but I think you might be the only one who can answer this question. :)

**Comment by blakeNaccarato** [comment]
Basically, `Model.model_fields` maps field names to their `FieldInfo` instances, each with a `FieldInfo.annotation` attribute roughly representing the same type information that `type_` gave before. However, that annotation could be a proper instance of `type`, or `typing.Annotated` (could be nested, even), or `types.GenericAlias`, or a container like `tuple` or `dict` composed of all of the above, and on and on. So you'll need to write a handler that can process a set of annotations you expect to encounter.

You end up needing to process that `FieldInfo.annotation` value by using the `typing` module's `get_args` and `get_origin` (import from `typing_extensions` in earlier Python versions) to process type annotations into concrete types. `get_origin` will unwrap unwieldy `GenericAlias` instances, and `get_args` will return tuples of types inside of generic annotations (e.g. `dict[str, float]`).

The Pydantic v2 docs on [handling custom generic classes](https://docs.pydantic.dev/latest/concepts/types/#handling-custom-generic-classes) touches on `get_args` and `get_origin` for a start, but yes it's generally non-trivial to write custom type handling code!

For example, see below for a handler I wrote to get types out of `collections.abc.Mapping/Sequence` generic annotations as well as "scalar" types. You might use it like `[tuple(get_types(field.annotation)) for field in Model.model_fields.values()]`. I wrote it as an iterator to simplify the recursive calls. It is for my specific needs over in https://github.com/softboiler/boilercore/blob/main/src/boilercore/__init__.py.



```Python
from collections.abc import Iterator, Mapping, Sequence
from itertools import chain
from types import EllipsisType, GenericAlias
from typing import Annotated, get_args, get_origin


def get_types(
    annotation: type | EllipsisType | GenericAlias,
) -> Iterator[type | EllipsisType]:
    """Get types for scalar, mapping, sequence, and annotated types."""
    if isinstance(annotation, type | EllipsisType):
        yield annotation
    elif (origin := get_origin(annotation)) and (args := get_args(annotation)):
        if issubclass(origin, Mapping):
            value_annotation = args[-1]
            yield from get_types(value_annotation)
        elif issubclass(origin, Sequence):
            yield from chain.from_iterable(get_types(a) for a in args)
        elif issubclass(origin, Annotated):
            annotated_type = args[0]
            yield from get_types(annotated_type)
    else:
        raise TypeError("Type not supported.")

```
